### PR TITLE
Simplify Exists

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: src/main/haskell/kore/src/Kore/Step/Simplification/Exists.hs [hunk 1]
```diff
@@ -9,79 +9,293 @@ Portability : portable
...
+The simplification of exists x . (pat and pred and subst) is equivalent to:
+
+* If the subst contains an assignment for x, then substitute that in pat and
+  pred, reevaluate them and return
+  (reevaluated-pat and reevaluated-pred and subst-without-x).
+* Otherwise, if x does not occur free in pat and pred, return
+  (pat and pred and subst)
+* Otherwise, if x does not occur free in pat, return
+  (pat and (exists x . pred) and subst)
+* Otherwise, if x does not occur free in pat, return
```
Comment (traiansf) [inline src/main/haskell/kore/src/Kore/Step/Simplification/Exists.hs]: I think you want to say "does not occur free in *pred*" here.

File: src/main/haskell/kore/test/Test/Kore/Step/Simplification/Exists.hs [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            )
+            (evaluate mockMetadataTools
+                (makeExists
+                    Mock.x
+                    [something1OfXExpanded, something2OfXExpanded]
+                )
+            )
+        )
+    , testCase "Exists - bool operations"
+        (do
```
Comment (traiansf) [inline src/main/haskell/kore/test/Test/Kore/Step/Simplification/Exists.hs]: Why are these in a do?

Comment (virgil-serbanuta) [reply src/main/haskell/kore/test/Test/Kore/Step/Simplification/Exists.hs]: It seemed to make sense to group them together, since they were fairly similar

**Comment by virgil-serbanuta** [comment]
Jenkins: test this please

**Pull request merged**